oneoffs/joseki/src/App.js

```diff
@@ -0,0 +1,455 @@
+});
+
+const colors = colormap({
+  colormap: 'copper',
+  nshades: NUM_COLORS+1,
+  format: 'rgbaString',
+  alpha: [0.02,1]
+});
+
+const theme = createMuiTheme({
+  palette: {
+    primary: { main: '#1976d2' },
+    secondary: { main: '#D32F2F' }
+  }
+});
+
+
+// Lol "const"
+const defaultHighlights = {}
+defaultHighlights[colors[5]] = flatten(range(10).map(idx => {
+  return range(10-idx).map(jdx => {
+    return {x: 18-idx, y:9-jdx };
+  });
+}));
+
+class DefaultDict {
+  constructor(defaultInit) {
+    return new Proxy({}, {
+      get: (target, name) => name in target ?
+        target[name] :
+        (target[name] = typeof defaultInit === 'function' ?
+          new defaultInit().valueOf() :
+          defaultInit)
+      })
+  }
+}
+
+const topLeft = {x: 9, y: 0};
+const bottomRight = {x: 18, y: 9};
+
+class Joseki extends React.Component {
+  constructor(props) {
+        super(props);
+        this.state = {
+          moves: "",
+          board: new godash.Board(),
+          nextColor: godash.BLACK,
+          passNext: true,  // was 'tenuki' one of the next moves played here.
+          other_highlights: defaultHighlights, // if so, what moves were considered then?
+          highlights: defaultHighlights,
+          search_enabled: false,
+          chartData: null,
+          tableData: null,
+          count: null,
+          run: null,
+          tableHourSort: 'desc',
+          tablePage: 1,
+          doneMessage: false,
+        };
+
+        this.resetBoard = this.resetBoard.bind(this);
+        this.coordinateClicked = this.coordinateClicked.bind(this);
+        this.updateGraph = this.updateGraph.bind(this);
+        this.findGames = this.findGames.bind(this);
+        this.tenuki = this.tenuki.bind(this);
+        this.handleRunChange = this.handleRunChange.bind(this);
+        this.prevMove = this.prevMove.bind(this);
+        this.handleKeyDown = this.handleKeyDown.bind(this);
+        this.toggleHourSort = this.toggleHourSort.bind(this);
+        this.handleNextButtonClick = this.handleNextButtonClick.bind(this);
+        this.handlePrevButtonClick = this.handlePrevButtonClick.bind(this);
+
+        this.updateHeatmap(); // update our heatmap with the empty board.
+}
+
+  handleKeyDown(key, e) {
+    if (key === 'left') {
+      this.prevMove(e);
+    }
+  }
+
+  handleNextButtonClick() {
+    this.setState(
+        {tablePage: this.state.tablePage + 1},
+        () => {this.findGames();}
+    );
+  }
+  handlePrevButtonClick() {
+    this.setState(
+        {tablePage: this.state.tablePage - 1},
+        () => {this.findGames();}
+    );
+  }
+
+  resetBoard() {
+    this.setState({
+      board: new godash.Board(),
+      nextColor: godash.BLACK,
+      passNext: true,
+      moves: "",
+      highlights: null,
+      other_highlights: null,
+      chartData: null,
+      tableData: null,
+      tablePage: 1,
+      search_enabled: false,
+      run: null,
+      doneMessage: false,
+    }, () => {
```

Comment:
is the "() => { this.updateHeatmap() }" a callback on success of setState?

Reply:
yes.